File: TS/hellointerview/two-pointers/sort-colors.ts

class Solution {
  sortColors(nums: number[]): void {
    let left = 0;
    let right = nums.length - 1;

    let i = 0;

    while (i < right) {
      if (nums[i] === 2) {
        [nums[i], nums[right]] = [nums[right], nums[i]];
        right--;
      } else if (nums[i] === 0) {
        [nums[i], nums[left]] = [nums[left], nums[i]];
        left++;
        i++;
      } else {
        i++;
      }
    }
  }
}

File: Java/hellointerview/01-two-pointers/sortColors.java
class Solution {
  public void sortColors(int[] nums) {
    int left = 0, right = nums.length - 1;
    int i = 0;

    while (i <= right) {
      if (nums[i] == 0) {
        int temp = nums[i];
        nums[i] = nums[left];
        nums[left] = temp;
        left++;
        i++;
      } else if (nums[i] == 2) {
        int temp = nums[i];
        nums[i] = nums[right];
        nums[right] = temp;
        right--;
      } else {
        i++;
      }
    }
  }
}
